File: python/leap/leap.py
__LEAP_YEAR: int = 4
__CENTURIAL_YEAR: int = 100
__LEAP_CYCLE_LENGTH: int = __LEAP_YEAR * __CENTURIAL_YEAR


def leap_year(year: int) -> bool:
    return (
        __is_leap_year(year) and
        (__is_non_centurial_year(year) or __is_leap_cycle_year(year))
    )


def __is_leap_year(year: int) -> bool:
    return year % __LEAP_YEAR == 0


def __is_non_centurial_year(year: int) -> bool:
    return not year % __CENTURIAL_YEAR == 0


def __is_leap_cycle_year(year: int) -> bool:
    return year % __LEAP_CYCLE_LENGTH == 0

File: cpp/leap/leap.cpp
#include "leap.h"

namespace leap {
  bool is_leap_year(int year) {
    using namespace detail;

    return leap_year(year) &&
      (is_non_centurial_year(year) || is_leap_year_cycle(year));
  }

  namespace detail {
    bool leap_year(int year) {
      return year % LEAP_YEAR == 0;
    }

    bool is_non_centurial_year(int year) {
      return !(year % CENTURIAL_YEAR == 0);
    }

    bool is_leap_year_cycle(int year) {
      return year % LEAP_CYCLE_LENGTH == 0;
    }
  }
}
